### ISSUE TITLE
Make it easier to add extra information to cart / order lines

### ISSUE THREAD

**Issue by pbright** [sonsandco/django-shoptools#4]
If you want to add some information to a cart / order line, currently there's a bunch of init methods etc that you have to edit to do so. Might not be possible but it would be nice to make this easier.

Two situations I'm thinking of here:
1. Additional notes that may be added to a line, such as monogramming or gift card details
2. Additional information that is required for site specific implementation, such as an information required by the shipping module.

In the case of an OrderLine, this would somehow all need to be saved to the line itself (rather than using any fk) to avoid errors upon a product subsequently being deleted.

**Comment by gregplaysguitar** [comment]
@pbright, have a look at b51ff5bcf64283a4d5ef25dfe9fdc3ffc40414cf - does that solve the problem for your use case? Options are defined by the product (`available_options`) and saved to the OrderLine as a postgres JSONField

**Comment by pbright** [comment]
This seems good for most use cases.

The use case I was thinking of for point 2 is where a product counts as more than one thing for shipping purposes - so some "product" that is a composite. For example a product which is a case of wine might need to count as 12 items for shipping (since they'd make up cases or part cases from products that are individual bottles, so the shipping cost is going to depend on total bottles).

The amount of things it counts as for shipping needs to be saved to the line in case the product is subsequently deleted.

It could be easily done using this approach by putting an option for it (with only one value) in available_options, but you might not be happy with using that terminology for it? I'm pretty comfortable with it as is because I think this is going to be an infrequently encountered use case.

**Comment by pbright** [comment]
Currently looks like there's no easy way to give these prices? Obviously it could be calculated using code but then that gets really messy if the cost for an option is changed, so it would be preferable to save it to the line.

This could be solved by having a tuple as the value instead of a string; so I think this will be fixed once the TODO comment about not requiring them to be strings is completed. Does that sound right?

**Comment by gregplaysguitar** [comment]
Hmm, not sure if I follow, but I think you just need to approach it in a slightly different way

> For example a product which is a case of wine might need to count as 12 items for shipping

In this case why can't your shipping module examine the cart/order contents and determine what it costs based on the line items? I don't see why it needs to store an option on the line.

> The amount of things it counts as for shipping needs to be saved to the line in case the product is subsequently deleted.

I don't think this matters, because shipping_cost is only calculated on the fly for carts; it's stored on the order in the shipping_cost field once the order is confirmed and not recalculated (which is important, because if the shipping cost rules are changed we need it to stay the same for past orders)

**Comment by pbright** [comment]
Yes, you are right. Not sure why I thought point 2 was an issue.

**Comment by gregplaysguitar** [comment]
If you're happy, I'll close this?

**Comment by pbright** [comment]
Just started working on Wallrus again, and it reminded me that it was the use case I was thinking of. 

They have some products which are murals and the customer can chose how to crop them. So I need to pass through coordinates for the corners of the cropped area, both to display to the customer but also so the client knows how to alter the mural before sending it to the printer.

I can pass these through in the options dict and given it's probably an uncommon case I'm comfortable with that, but thought I should mention it as I assume it's not really what you intended options to be used for.

Also did you see my comment above re giving options prices?

**Comment by pbright** [comment]
Actually I can't do this using options as it will be stripped out by validate_options.

**Comment by pbright** [comment]
@gregplaysguitar 

I deleted a couple of my comment regarding a use case that after thinking about it further I don't think is appropriate to use here.

So I think what you've done is good. The only outstanding thing was my comment above regarding pricing for options.

**Comment by gregplaysguitar** [comment]
Still not sure if I understand the pricing issue. Might need you to talk me through it
